### ISSUE TITLE
Lots of "Stateless" React Components in no-multi-comp

### ISSUE THREAD

**Issue by epmatsw** [yannickcr/eslint-plugin-react#290]
After upgrading to 3.7.1, I've got a few new files that are showing `react/no-multi-comp` errors. That's because stuff like:

```
renderCell: function(item, index) {
      return React.createElement("td", {
        "key": index
      }, item.title);
    }
```
is picked up as a stateless component. While I could go through the few files that have this pattern and add exceptions to the rule, that seems unpleasant.

Would it be possible to take one of the following approaches?
1. Only count functions that take a single argument or a destructured single argument as components.
2. Add an option to `react/no-multi-comp` to exclude stateless components from the count.

**Comment by yannickcr** [comment]
Can you give me the full source code for a file that trigger `no-multi-comp` errors in your configuration ?

> 1. Only count functions that take a single argument or a destructured single argument as components.

Stateless components can also have 2 arguments (`props` and `context`).
> 2. Add an option to `react/no-multi-comp` to exclude stateless components from the count.

Yes, this is possible.

**Comment by epmatsw** [comment]
We're still using ES5-style classes (compiled Coffeescript 😳), so that's probably something to do with this. The offending code looks like:
```
Test.prototype.renderCell = function(item, index) {
        return React.createElement("td", {
          "key": index
        }, item.title);
      };

      Test.prototype.render = function() {
        return React.createElement("tr", null, this.props.items.map(this.renderCell));
      };
```

Full Gist: https://gist.github.com/epmatsw/350777c3731de0f77fba

**Comment by jsg2021** [comment]
I've seen many components split rendering into helper methods on the component... and I would not consider them to be "stateless" components... yet this rule is tripping up on those helper methods.

example, this is could be a mixin:

```es6
import React from 'react';

import HeaderWidget from './Header';
import FooterWidget from './Footer';

export default {

	renderHeader () {
		let {page} = this.state;
		
		return React.createElement(HeaderWidget, {
			//etc...,
			page
		});
	},


	renderFooter () {
		let {page} = this.state;
		if (!page) {
			return null;
		}

		return React.createElement(FooterWidget, {
			//etc...,
			page
		});
	}

};
```
ignore the stupidity of this example, but the idea is that methods either within the component or mixins are not themselves "stateless" components. they are just helpers.  They could have many or no arguments. 

in my opinion, it's too complicated to identify stateless components or helpers... so lint rules that were built with previous assumptions about components need re-evaluating...  

tl;dr: imo, the no-muli-comp rule should only apply to Stateful (traditional) Components.

**Comment by yannickcr** [comment]
> in my opinion, it's too complicated to identify stateless components or helpers... so lint rules that were built with previous assumptions about components need re-evaluating... 

Yes, that's what we are trying to do, but since it is really difficult to detect stateless components we try to make some assumptions and treat some exceptions. And the diversity of the patterns used in the React community is not helping ;)

> tl;dr: imo, the no-muli-comp rule should only apply to Stateful (traditional) Components.

I will add an option for this for sure but I don't plan to totally deactivate this rule for stateless components.
